components/nitro_utils/cose.cc

```diff
@@ -0,0 +1,253 @@
+          cbor::Reader::DecoderError::CBOR_NO_ERROR) {
+    LOG(ERROR) << "COSE: failed to read payload cbor: "
+               << cbor::Reader::ErrorCodeToString(*cbor_config.error_code_out);
+    return false;
+  }
+  DCHECK(payload_dec_val.has_value());
+  payload = payload_dec_val->Clone();
+
+  const cbor::Value& signature_val = cose_arr[3];
+  if (!signature_val.is_bytestring()) {
+    LOG(ERROR) << "COSE: signature value is not bstr";
+    return false;
+  }
+  signature = signature_val.GetBytestring();
+
+  return true;
+}
+
+bool CoseSign1::Verify(const net::ParsedCertificateList& cert_chain) {
+  DCHECK_GT(cert_chain.size(), 1U);
+
+  net::der::GeneralizedTime time_now;
+  DCHECK(net::der::EncodeTimeAsGeneralizedTime(base::Time::Now(), &time_now));
+
+  net::CertPathErrors cert_path_errors;
+
+  net::VerifyCertificateChain(
+      cert_chain, net::CertificateTrust::ForTrustAnchorEnforcingExpiration(),
+      this, time_now, net::KeyPurpose::ANY_EKU,
+      net::InitialExplicitPolicy::kFalse, std::set<net::der::Input>(),
+      net::InitialPolicyMappingInhibit::kFalse,
+      net::InitialAnyPolicyInhibit::kFalse, nullptr, &cert_path_errors);
+
+  if (cert_path_errors.ContainsHighSeverityErrors()) {
+    LOG(ERROR) << "COSE verification: bad certificate chain: "
+               << cert_path_errors.ToDebugString(cert_chain);
+    return false;
+  }
+
+  std::vector<cbor::Value> sig_data_vec;
+  sig_data_vec.emplace_back(cbor::Value("Signature1"));
+  sig_data_vec.push_back(protected_encoded.Clone());
+  sig_data_vec.emplace_back(cbor::Value(std::vector<uint8_t>()));
+  sig_data_vec.push_back(payload_encoded.Clone());
+  cbor::Value sig_data(sig_data_vec);
+
+  absl::optional<std::vector<uint8_t>> encoded_sig_data =
+      cbor::Writer::Write(sig_data);
+  DCHECK(encoded_sig_data.has_value());
+
+  base::StringPiece low_cert_spki;
+  if (!net::asn1::ExtractSPKIFromDERCert(
+          cert_chain.front()->der_cert().AsStringPiece(), &low_cert_spki)) {
+    LOG(ERROR) << "COSE verification: could not extract SPKI from cert";
+    return false;
+  }
+
+  std::vector<uint8_t> low_cert_spki_vec;
+  std::transform(low_cert_spki.begin(), low_cert_spki.end(),
+                 std::back_inserter(low_cert_spki_vec),
+                 [](char c) { return static_cast<uint8_t>(c); });
+
+  std::vector<uint8_t> sig_der;
+  if (!ConvertCoseSignatureToDER(signature, &sig_der)) {
+    return false;
+  }
+
+  crypto::SignatureVerifier sig_verifier;
+  if (!sig_verifier.VerifyInit(
+          crypto::SignatureVerifier::SignatureAlgorithm::ECDSA_SHA384, sig_der,
+          low_cert_spki_vec)) {
+    LOG(ERROR) << "COSE verification: failed to init cert verifier";
+    return false;
+  }
+
+  sig_verifier.VerifyUpdate(*encoded_sig_data);
+
+  return sig_verifier.VerifyFinal();
+}
+
+bool CoseSign1::IsSignatureAlgorithmAcceptable(
+    const net::SignatureAlgorithm& signature_algorithm,
+    net::CertErrors* errors) {
+  return true;
```

Comment:
The certificate chain verification function requires a `net::VerifyCertificateChainDelegate`, so `IsSignatureAlgorithmAcceptable` and `IsPublicKeyAcceptable` were implemented to fulfill this abstract class. Since we're not too picky about the signature algorithm or specific public key, and only concerned about a valid cert chain, we return true for these methods.